### PR TITLE
Poseidon MWE

### PR DESCRIPTION
**Pull request by mpenciak**
This PR implements a basic version of the Poseidon hash function.

As of now the hash implementation passes the `poseidonperm_x5_255_5 ` test in the [reference implementation](https://extgit.iaik.tugraz.at/krypto/hadeshash). 

TODO List before merging:

- [x] Add the Lurk `Poseidon.Profile`
- [ ] Write Poseidon hash tests for Lurk profile
- [x] Move the MDS and Round Constants to a separate "Constants"
- [ ] Documentation
- [x] Add license

### CODE REVIEW

File: Poseidon/ForYatimaStdLib.lean [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+-- Shifts the byte array left by 1, preserves length (so in particular kills the first coefficient
+def ByteArray.shiftLeft (bs : ByteArray) : ByteArray := Id.run do
+  let mut answer : ByteArray := .mkEmpty bs.size
+  for idx in [:bs.size] do
+    answer := answer.push <| (bs[idx]! <<< 1) + (bs.getD (idx + 1) >>> 7)
+  answer
+
+def ByteArray.shiftAdd (bs : ByteArray) (b : Bit) : ByteArray :=
+  let ans := bs.shiftLeft
+  ans.set! (ans.size - 1) (ans[(ans.size - 1)]! + b.toUInt8)
```
Comment (DanielRrr) [inline Poseidon/ForYatimaStdLib.lean:107]: All those definitions are in the standard library now, so consider bumping `YatimaStdLib`

File: Poseidon/Tests/MDS.lean [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import LSpec
+import Poseidon.Parameters.PoseidonPerm255
+
+open LSpec
+
```
Comment (DanielRrr) [inline Poseidon/Tests/MDS.lean:5]: Is it in progress now?

**Comment by mpenciak** [comment]
This PR also closes #9.

What's missing is documentation and LSpec tests, I'll make separate PRs for that so we're not holding up its use in `Lurk.lean`

**Review by arthurpaulino — CHANGES_REQUESTED** [review]

File: Poseidon/Tests/Hash.lean [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+def output : Vector (Zmod testProfile.p) := 
+             #[0x2a918b9c9f9bd7bb509331c81e297b5707f6fc7393dcee1b13901a0b22202e18,
+               0x65ebf8671739eeb11fb217f2d5c5bf4a0c3f210e3f3cd3b08b5db75675d797f7,
+               0x2cc176fc26bc70737a696a9dfd1b636ce360ee76926d182390cdb7459cf585ce,
+               0x4dc4e29d283afd2a491fe6aef122b9a968e74eff05341f3cc23fda1781dcb566,
+               0x03ff622da276830b9451b88b85e6184fd6ae15c8ab3ee25a5667be8592cce3b1]
+
+#lspec 
+  test "poseidonperm_x5_255_5" (Poseidon.hashInput testProfile testContext input == output)
```
[the remark this answers is not in the document]
Comment (arthurpaulino) [reply Poseidon/Tests/Hash.lean:33]: That would allow you to call `lake exe lspec` to run the test suite

[the remark this answers is not in the document]
Comment (mpenciak) [reply Poseidon/Tests/Hash.lean:33]: Yeah the main testing suite is going to be setup that way with a github CI integration, the test was just in there for personal use while I was developing to make sure I got everything right.

[the remark this answers is not in the document]
Comment (arthurpaulino) [reply Poseidon/Tests/Hash.lean:33]: Ah so it's a WIP

**Review by arthurpaulino — APPROVED** [review]

**Pull request merged**